File: 199-binary-tree-right-side-view/solution.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

type Traverse = { node: TreeNode; depth: number };

function rightSideView(root: TreeNode | null): number[] {
  if (!root) return [];
  const answer: number[] = [];
  const traversal: Traverse[] = [{ node: root, depth: 0 }];
  while (traversal[0]) {
    const { node, depth } = traversal.shift() as Traverse;
    answer[depth] = node.val;
    if (node.left) traversal.push({ node: node.left, depth: depth + 1 });
    if (node.right) traversal.push({ node: node.right, depth: depth + 1 });
  }
  return answer;
}

File: 199-binary-tree-right-side-view/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void traversal(vector<vector<int>>& vals, int depth, TreeNode* node) {
        if(node == NULL) return;
        if(depth >= vals.size()) vals.push_back({});
        vals[depth].push_back(node->val);
        traversal(vals, depth + 1, node->left);
        traversal(vals, depth + 1, node->right);
    }

    vector<int> rightSideView(TreeNode* root) {
        vector<vector<int>> vals;
        traversal(vals, 0, root);
        vector<int> ans;
        for(auto x : vals) ans.push_back(x[x.size() - 1]);
        return ans;        
    }
};
